File: data_prep/gdp_cap.py
# This file reads gdp / cap data from three sources (WDI, maddison, PWT),
# formats it and saves it. 

import pandas as pd

################
### WDI data ###
################

# WDI data is from: https://databank.worldbank.org/source/world-development-indicators#
# Measured in constant 2011 intl dollars (PPP)

# read data, skipping first 4 lines (download date and empty)
wdi = pd.read_csv('~/sorensfolder/sbpdata/data/gdp/wdi/wdi_gdp_cap_2011ppp.csv')

# remove spaces from column names
wdi.columns = wdi.columns.str.replace(' ', '')

# change names
wdi.rename(columns={'CountryName':'country',
                    'CountryCode':'code'},
           inplace=True)

# drop indicator cols
wdi = wdi.drop(columns=['SeriesName', 'SeriesCode'])

# melt to tidy
wdi_tidy = pd.melt(wdi,
                   id_vars=['country', 'code'],
                   var_name='year',
                   value_name='gdp_cap')

# remove code from year col, convert to int
wdi_tidy.loc[:, 'year'].replace(regex=True,
                         inplace=True,
                         to_replace='\[YR....\]',
                         value='')

# change to correct col types
wdi_tidy = wdi_tidy.astype({'year':int,
                            'gdp_cap':numeric})


# write the file
wdi_tidy.to_csv('../data/gdp/WDI_gdp_cap_ppp2011_tidy.csv')


# TODO NOT DONE, REDOWNLOAD DATA WITHOUT CODES; MISSING AS ' '


#####################
### Maddison data ###
#####################

# Maddison data is from: https://www.rug.nl/ggdc/historicaldevelopment/maddison/

File: data_prep/gdp_cap.R
# This file reads gdp / cap data from three sources (WDI, maddison, PWT), formats and saves it.
library(tidyverse)
library(haven)

# pwt9 contains the penn world tables v9
library(pwt9)
# wdi contains api for wdi
library(WDI)
# maddison conatains the maddison database, updated 2018
# install_github("expersso/maddison")
library(maddison)

## OUTPUT PATH
output_dir <- "../data/joined_gdp_cap.csv"

################
### WDI data ###
################
# WDI data is from: https://databank.worldbank.org/source/world-development-indicators#
# Measured in constant 2011 intl dollars (PPP)

wdi <- as_tibble(WDI(country = "all", 
		      indicator = "NY.GDP.PCAP.PP.CD",
		      extra = TRUE,
		      start = NULL))

wdi_tidy <- wdi %>%
	select(country = iso3c, year, wdi_gdp_cap = NY.GDP.PCAP.PP.CD) %>%
	mutate_if(is.factor, as.character)

#####################
### Maddison data ###
#####################

# Maddison data is from: https://www.rug.nl/ggdc/historicaldevelopment/maddison/
mad <- as_tibble(maddison)

# for computing relative growth rates, authors recommend the RGDPNApc
mad_tidy <- mad %>%
	select(country = iso3c, year, mad_gdp_cap = rgdpnapc)

################
### PWT data ###
################

# pwt data is from https://www.rug.nl/ggdc/productivity/pwt
pwt_tidy <- as_tibble(pwt9.1) %>%
	select(country = isocode, year, rgdpe, pop) %>% 
	mutate(pwt_gdp_cap = rgdpe / pop) %>% 
	mutate_if(is.factor, as.character) %>%
	select(country, year, pwt_gdp_cap)

#######################
### join the frames ###
#######################
joined <- wdi_tidy %>%
	left_join(mad_tidy) %>%
	left_join(pwt_tidy) 

write_csv(joined, output_dir)
